// @rsksmart/rns/lib/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ADDR_REVERSE_NAMEHASH = exports.AVAILABLE_TLDS = exports.AVAILABLE_INTERFACE = exports.SET_NAME_INTERFACE = exports.NAME_INTERFACE = exports.CONTENTHASH_INTERFACE = exports.NEW_ADDR_INTERFACE = exports.SET_CHAIN_ADDR_INTERFACE = exports.CHAIN_ADDR_INTERFACE = exports.SET_ADDR_INTERFACE = exports.ADDR_INTERFACE = exports.DEVPORTAL_ERRORS_URL = exports.ZERO_ADDRESS = void 0;
var ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';
exports.ZERO_ADDRESS = ZERO_ADDRESS;
var DEVPORTAL_ERRORS_URL = 'https://developers.rsk.co/rif/rns/libs/javascript/Errors/'; // contracts interfaces

exports.DEVPORTAL_ERRORS_URL = DEVPORTAL_ERRORS_URL;
var ADDR_INTERFACE = '0x3b3b57de';
exports.ADDR_INTERFACE = ADDR_INTERFACE;
var SET_ADDR_INTERFACE = '0xd5fa2b00';
exports.SET_ADDR_INTERFACE = SET_ADDR_INTERFACE;
var CHAIN_ADDR_INTERFACE = '0x8be4b5f6';
exports.CHAIN_ADDR_INTERFACE = CHAIN_ADDR_INTERFACE;
var SET_CHAIN_ADDR_INTERFACE = '0xd278b400';
exports.SET_CHAIN_ADDR_INTERFACE = SET_CHAIN_ADDR_INTERFACE;
var NEW_ADDR_INTERFACE = '0xf1cb7e06';
exports.NEW_ADDR_INTERFACE = NEW_ADDR_INTERFACE;
var CONTENTHASH_INTERFACE = '0xbc1c58d1';
exports.CONTENTHASH_INTERFACE = CONTENTHASH_INTERFACE;
var NAME_INTERFACE = '0x691f3431';
exports.NAME_INTERFACE = NAME_INTERFACE;
var SET_NAME_INTERFACE = '0xc47f0027';
exports.SET_NAME_INTERFACE = SET_NAME_INTERFACE;
var AVAILABLE_INTERFACE = '0x96e494e8';
exports.AVAILABLE_INTERFACE = AVAILABLE_INTERFACE;
var AVAILABLE_TLDS = ['rsk'];
exports.AVAILABLE_TLDS = AVAILABLE_TLDS;
var ADDR_REVERSE_NAMEHASH = '0x91d1777781884d03a6757a803996e38de2a42967fb37eeaca72729271025a9e2';
exports.ADDR_REVERSE_NAMEHASH = ADDR_REVERSE_NAMEHASH;
