// @rsksmart/rns/lib/errors/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ErrorWrapper", {
  enumerable: true,
  get: function get() {
    return _ErrorWrapper.default;
  }
});
exports.CHAIN_ID_NOT_SUPPORTED = exports.NO_CONTENTHASH_SET = exports.UNSUPPORTED_CONTENTHASH_PROTOCOL = exports.NO_CONTENTHASH_INTERFACE = exports.NO_SET_CHAIN_ADDR = exports.NO_SET_NAME_METHOD = exports.NO_REVERSE_REGISTRAR = exports.NO_AVAILABLE_METHOD = exports.NO_TLD_OWNER = exports.INVALID_CHECKSUM_ADDRESS = exports.NO_SET_ADDR = exports.INVALID_ADDRESS = exports.SUBDOMAIN_NOT_AVAILABLE = exports.NO_ACCOUNTS_TO_SIGN = exports.NO_REVERSE_RESOLUTION_SET = exports.NO_NAME_RESOLUTION = exports.DOMAIN_NOT_EXISTS = exports.INVALID_LABEL = exports.INVALID_DOMAIN = exports.SEARCH_DOMAINS_UNDER_AVAILABLE_TLDS = exports.SEARCH_ONLY_SIMPLE_DOMAINS = exports.NO_CHAIN_ADDR_RESOLUTION_SET = exports.NO_CHAIN_ADDR_RESOLUTION = exports.NO_ADDRESSES_PROVIDED = exports.LIBRARY_NOT_COMPOSED = exports.NO_RESOLVER = exports.NO_ADDR_RESOLUTION = exports.NO_ADDR_RESOLUTION_SET = exports.default = void 0;

var _RNSError = _interopRequireDefault(require("./RNSError"));

var _ErrorWrapper = _interopRequireDefault(require("./ErrorWrapper"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = _RNSError.default;
exports.default = _default;
var NO_ADDR_RESOLUTION_SET = 'NO_ADDR_RESOLUTION_SET';
exports.NO_ADDR_RESOLUTION_SET = NO_ADDR_RESOLUTION_SET;
var NO_ADDR_RESOLUTION = 'NO_ADDR_RESOLUTION';
exports.NO_ADDR_RESOLUTION = NO_ADDR_RESOLUTION;
var NO_RESOLVER = 'NO_RESOLVER';
exports.NO_RESOLVER = NO_RESOLVER;
var LIBRARY_NOT_COMPOSED = 'LIBRARY_NOT_COMPOSED';
exports.LIBRARY_NOT_COMPOSED = LIBRARY_NOT_COMPOSED;
var NO_ADDRESSES_PROVIDED = 'NO_ADDRESSES_PROVIDED';
exports.NO_ADDRESSES_PROVIDED = NO_ADDRESSES_PROVIDED;
var NO_CHAIN_ADDR_RESOLUTION = 'NO_CHAIN_ADDR_RESOLUTION';
exports.NO_CHAIN_ADDR_RESOLUTION = NO_CHAIN_ADDR_RESOLUTION;
var NO_CHAIN_ADDR_RESOLUTION_SET = 'NO_CHAIN_ADDR_RESOLUTION_SET';
exports.NO_CHAIN_ADDR_RESOLUTION_SET = NO_CHAIN_ADDR_RESOLUTION_SET;
var SEARCH_ONLY_SIMPLE_DOMAINS = 'SEARCH_ONLY_SIMPLE_DOMAINS';
exports.SEARCH_ONLY_SIMPLE_DOMAINS = SEARCH_ONLY_SIMPLE_DOMAINS;
var SEARCH_DOMAINS_UNDER_AVAILABLE_TLDS = 'SEARCH_DOMAINS_UNDER_AVAILABLE_TLDS';
exports.SEARCH_DOMAINS_UNDER_AVAILABLE_TLDS = SEARCH_DOMAINS_UNDER_AVAILABLE_TLDS;
var INVALID_DOMAIN = 'INVALID_DOMAIN';
exports.INVALID_DOMAIN = INVALID_DOMAIN;
var INVALID_LABEL = 'INVALID_LABEL';
exports.INVALID_LABEL = INVALID_LABEL;
var DOMAIN_NOT_EXISTS = 'DOMAIN_NOT_EXISTS';
exports.DOMAIN_NOT_EXISTS = DOMAIN_NOT_EXISTS;
var NO_NAME_RESOLUTION = 'NO_NAME_RESOLUTION';
exports.NO_NAME_RESOLUTION = NO_NAME_RESOLUTION;
var NO_REVERSE_RESOLUTION_SET = 'NO_REVERSE_RESOLUTION_SET';
exports.NO_REVERSE_RESOLUTION_SET = NO_REVERSE_RESOLUTION_SET;
var NO_ACCOUNTS_TO_SIGN = 'NO_ACCOUNTS_TO_SIGN';
exports.NO_ACCOUNTS_TO_SIGN = NO_ACCOUNTS_TO_SIGN;
var SUBDOMAIN_NOT_AVAILABLE = 'SUBDOMAIN_NOT_AVAILABLE';
exports.SUBDOMAIN_NOT_AVAILABLE = SUBDOMAIN_NOT_AVAILABLE;
var INVALID_ADDRESS = 'INVALID_ADDRESS';
exports.INVALID_ADDRESS = INVALID_ADDRESS;
var NO_SET_ADDR = 'NO_SET_ADDR';
exports.NO_SET_ADDR = NO_SET_ADDR;
var INVALID_CHECKSUM_ADDRESS = 'INVALID_CHECKSUM_ADDRESS';
exports.INVALID_CHECKSUM_ADDRESS = INVALID_CHECKSUM_ADDRESS;
var NO_TLD_OWNER = 'NO_TLD_OWNER';
exports.NO_TLD_OWNER = NO_TLD_OWNER;
var NO_AVAILABLE_METHOD = 'NO_AVAILABLE_METHOD';
exports.NO_AVAILABLE_METHOD = NO_AVAILABLE_METHOD;
var NO_REVERSE_REGISTRAR = 'NO_REVERSE_REGISTRAR';
exports.NO_REVERSE_REGISTRAR = NO_REVERSE_REGISTRAR;
var NO_SET_NAME_METHOD = 'NO_SET_NAME_METHOD';
exports.NO_SET_NAME_METHOD = NO_SET_NAME_METHOD;
var NO_SET_CHAIN_ADDR = 'NO_SET_CHAIN_ADDR';
exports.NO_SET_CHAIN_ADDR = NO_SET_CHAIN_ADDR;
var NO_CONTENTHASH_INTERFACE = 'NO_CONTENTHASH_INTERFACE';
exports.NO_CONTENTHASH_INTERFACE = NO_CONTENTHASH_INTERFACE;
var UNSUPPORTED_CONTENTHASH_PROTOCOL = 'UNSUPPORTED_CONTENTHASH_PROTOCOL';
exports.UNSUPPORTED_CONTENTHASH_PROTOCOL = UNSUPPORTED_CONTENTHASH_PROTOCOL;
var NO_CONTENTHASH_SET = 'NO_CONTENTHASH_SET';
exports.NO_CONTENTHASH_SET = NO_CONTENTHASH_SET;
var CHAIN_ID_NOT_SUPPORTED = 'CHAIN_ID_NOT_SUPPORTED';
exports.CHAIN_ID_NOT_SUPPORTED = CHAIN_ID_NOT_SUPPORTED;
