// @rsksmart/rns/lib/errors/RNSError.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _constants = require("../constants");

var _errors = _interopRequireDefault(require("./errors.json"));

var _enums = require("../types/enums");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class _default extends Error {
  constructor(errorId) {
    var lang = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : _enums.Lang.en;
    var error = _errors.default[errorId];

    if (!error) {
      error = {
        id: 'KB000',
        message: {
          en: 'Unknown error',
          es: 'Error desconocido',
          ja: '不明なエラー',
          ko: '알수없는 오류',
          pt: 'Erro desconhecido',
          ru: 'Неизвестная ошибка',
          zh: '未知错误'
        }
      };
    }

    super(error.message[lang]);
    this.id = error.id;
    this.ref = "".concat(_constants.DEVPORTAL_ERRORS_URL, "#").concat(error.id.toLowerCase());
  }

}

exports.default = _default;
module.exports = exports.default;
