// @rsksmart/rns/lib/types/enums.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Lang = exports.ChainId = exports.NetworkId = void 0;

/**
 * RSK Mainnet and Testnet network ids. Used to identify the current network.
 */
var NetworkId;
/**
 * Represents some of the chain hexas listed in SLIP44 (https://github.com/satoshilabs/slips/blob/master/slip-0044.md)
 */

exports.NetworkId = NetworkId;

(function (NetworkId) {
  NetworkId[NetworkId["RSK_MAINNET"] = 30] = "RSK_MAINNET";
  NetworkId[NetworkId["RSK_TESTNET"] = 31] = "RSK_TESTNET";
})(NetworkId || (exports.NetworkId = NetworkId = {}));

var ChainId;
/**
 * Supported feedback languages
 */

exports.ChainId = ChainId;

(function (ChainId) {
  ChainId["RSK"] = "0x80000089";
  ChainId["BTC"] = "0x80000000";
  ChainId["ETH"] = "0x8000003c";
  ChainId["LTC"] = "0x80000002";
  ChainId["DOGE"] = "0x80000003";
  ChainId["MONA"] = "0x80000016";
  ChainId["DASH"] = "0x80000005";
  ChainId["ETC"] = "0x8000003d";
  ChainId["XDAI"] = "0x800002bc";
  ChainId["XRP"] = "0x80000090";
  ChainId["BCH"] = "0x80000091";
  ChainId["BNB"] = "0x800002ca";
  ChainId["XLM"] = "0x80000094";
  ChainId["ATOM"] = "0x80000076";
  ChainId["TRX"] = "0x800000c3";
  ChainId["NEM"] = "0x8000002b";
  ChainId["EOS"] = "0x800000c2";
  ChainId["KSM"] = "0x800001b2";
})(ChainId || (exports.ChainId = ChainId = {}));

var Lang;
exports.Lang = Lang;

(function (Lang) {
  Lang["en"] = "en";
  Lang["es"] = "es";
  Lang["ja"] = "ja";
  Lang["ko"] = "ko";
  Lang["pt"] = "pt";
  Lang["ru"] = "ru";
  Lang["zh"] = "zh";
})(Lang || (exports.Lang = Lang = {}));
