// @rsksmart/rns/types/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ChainId", {
  enumerable: true,
  get: function get() {
    return _enums.ChainId;
  }
});
Object.defineProperty(exports, "NetworkId", {
  enumerable: true,
  get: function get() {
    return _enums.NetworkId;
  }
});
Object.defineProperty(exports, "RNS", {
  enumerable: true,
  get: function get() {
    return _rns.default;
  }
});
Object.defineProperty(exports, "Contracts", {
  enumerable: true,
  get: function get() {
    return _rns.Contracts;
  }
});
Object.defineProperty(exports, "ContractAddresses", {
  enumerable: true,
  get: function get() {
    return _options.ContractAddresses;
  }
});
Object.defineProperty(exports, "Options", {
  enumerable: true,
  get: function get() {
    return _options.Options;
  }
});
Object.defineProperty(exports, "Resolutions", {
  enumerable: true,
  get: function get() {
    return _resolutions.Resolutions;
  }
});
Object.defineProperty(exports, "Subdomains", {
  enumerable: true,
  get: function get() {
    return _subdomains.Subdomains;
  }
});
Object.defineProperty(exports, "Composable", {
  enumerable: true,
  get: function get() {
    return _composable.Composable;
  }
});
Object.defineProperty(exports, "Utils", {
  enumerable: true,
  get: function get() {
    return _utils.Utils;
  }
});
Object.defineProperty(exports, "Registrations", {
  enumerable: true,
  get: function get() {
    return _registrations.Registrations;
  }
});
Object.defineProperty(exports, "ErrorDictionary", {
  enumerable: true,
  get: function get() {
    return _errorDictionary.ErrorDictionary;
  }
});

var _enums = require("./enums");

var _rns = _interopRequireWildcard(require("./rns"));

var _options = require("./options");

var _resolutions = require("./resolutions");

var _subdomains = require("./subdomains");

var _composable = require("./composable");

var _utils = require("./utils");

var _registrations = require("./registrations");

var _errorDictionary = require("./error-dictionary");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
